rllib/agents/trainer.py

```diff
@@ -74,11 +74,26 @@
     # The dataflow here can vary per algorithm. For example, PPO further
     # divides the train batch into minibatches for multi-epoch SGD.
     "rollout_fragment_length": 200,
-    # Whether to rollout "complete_episodes" or "truncate_episodes" to
-    # `rollout_fragment_length` length unrolls. Episode truncation guarantees
-    # evenly sized batches, but increases variance as the reward-to-go will
-    # need to be estimated at truncation boundaries.
+    # How to build per-Sampler (RolloutWorker) batches, which are then
+    # usually concat'd to form the train batch. Note that "steps" below can
+    # mean different things (either env- or agent-steps) and depends on the
+    # `rollout_fragment_unit` setting below.
+    # truncate_episodes: Each produced batch (when calling
+    #   RolloutWorker.sample()) will contain exactly `rollout_fragment_length`
+    #   steps. This mode guarantees evenly sized batches, but increases
+    #   variance as the future return must now be estimated at truncation
+    #   boundaries.
+    # complete_episodes: Each unroll happens exactly over one episode, from
+    #   beginning to end. Data collection will not stop unless the episode
+    #   terminates or a configured horizon (hard or soft) is hit.
     "batch_mode": "truncate_episodes",
+    # Which metric to use as the "batch size" when building SampleBatch or
+    # MultiAgentBatch.
+    # env_steps: Count each time the env is "stepped" (no matter how many
+    #   multi-agent actions are passed/how many mult-agent observations have
+    #   been returned in the previous step).
+    # agent_steps: Count each individual agent step as one step.
+    "rollout_fragment_unit": "env_steps",
```

Comment:
How about `count_steps_by: env_steps`?

Reply:
I want to keep the connection to rollout_fragment: What about "count_fragments_by"?

Reply:
Doesn't it also affect train batch size, so in that case count_steps_by is more generic?

Reply:
As a user, prefer Sven's. It should not affect batch size -- that's the whole point of this change, to allow constant batch size (and hence memory) when your env has a variable number of agents. "rollout_fragment_unit" is perfect. Default to "env_steps", allow new option "agent_steps", and (only if it's trivial to implement, since this is niche) possibly "agent_rollouts"

Reply:
Ok, so one thing I want to check is that the unit of train batch size is always agent steps. Is this true?

Reply:
Checking https://github.com/ray-project/ray/blob/9ad0f173d6fbe00d41adad56ade0f40599390a87/rllib/execution/rollout_ops.py#L137

I don't believe train batch size is by agent steps, it is actually env steps. @sven1977 is this option supposed to apply to the train batch as well?

Reply:
Good point, if we are already doing train_batches by env_steps, we should extend this setting to train_batches as well. I'll confirm and add this.

Reply:
So we could do both:
a) Only count the Batches produced by the sampler in agent-steps (or env-steps by default)
b) Also count the train batches (produced by the ConcatBatches execution callable) in agent-steps (or env-steps by default).

I'll let you decide what's best. But we should - in either case - clearly specify, what the unit change would affect (either rollout-batches AND train-batch or only rollout-batches).

Also, in case b), we would need to change ConcatBatches so it knows what to count and stop concating once the train_batch_size is reached.

Reply:
@jsuarez5341 @ericl

Reply:
Is there a use case where you would only control one or the other? It seems more intuitive to change the units uniformly for both.

Reply:
Yeah, I agree. The only thing is that we then also have to change ConcatBatches (add another `unit` arg to its constructor so it knows how to count).

Reply:
Can you make the change to ConcatBatches then?